modin/pandas/groupby.py

```diff
@@ -171,10 +173,77 @@ def idxmax(self):
     def ndim(self):
         return 2  # ndim is always 2 for DataFrames
 
-    def shift(self, periods=1, freq=None, axis=0):
-        return self._default_to_pandas(
-            lambda df: df.shift(periods=periods, freq=freq, axis=axis)
-        )
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by periods observations.
+
+        If freq is passed, the index will be increased using the periods and the freq.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods to shift.
+        freq : str, optional
+            Frequency string.
+        axis : axis to shift, default 0
+            Shift direction.
+        fill_value : optional
+            The scalar value to use for newly introduced missing values.
+
+        Returns
+        -------
+        Series or DataFrame
+            Object shifted within each group.
+        """
+
+        def _shift(periods, freq, axis, fill_value, is_set_nan_rows=True):
+            from .dataframe import DataFrame
+
+            result = self._df.shift(periods, freq, axis, fill_value)
+
+            if is_set_nan_rows:
+                if (
+                    isinstance(self._by, BaseQueryCompiler)
+                    and set(self._by.columns).issubset(list(self._df.columns))
+                    and self._by.isna().to_numpy().any()
+                ):
+                    mask_nan_rows = self._df[self._by.columns].isna()
+                    if (isinstance(mask_nan_rows, DataFrame)) and len(
+                        mask_nan_rows.columns
+                    ) == 1:
+                        mask_nan_rows = mask_nan_rows.squeeze(axis=1)
+                    idx_nan_rows = mask_nan_rows[
+                        mask_nan_rows.any(axis=1)
+                        if (isinstance(mask_nan_rows, DataFrame))
+                        else mask_nan_rows
+                    ].index
+                    result.loc[idx_nan_rows] = np.nan
+            return result
+
+        if freq is None and axis == 1 and self._axis == 0 and freq is None:
+            result = _shift(periods, freq, axis, fill_value)
+        elif (
+            freq is not None
+            and axis == 0
+            and self._axis == 0
+            and isinstance(self._by, BaseQueryCompiler)
+        ):
+            result = _shift(periods, freq, axis, fill_value, is_set_nan_rows=False)
+            new_idx_lvl_arrays = [
+                self._df.loc[:, [col]].squeeze(axis=1).values
+                for col in self._by.columns
+            ] + [list(result.index)]
```

Comment:
grabbing columns by one in a for loop seems to be slightly slower, than grabbing all columns in a single query:
```
grabbing cols by one 0.6766219809651375
grabbing all 0.31146624498069286
```


<details><summary> Script to measure </summary>


```python
import numpy as np
import os
import pandas
from timeit import default_timer as timer


def generate_data_file(
    nrows=200000,
    ncols=256,
    filename=None,
    multiline_rows=False,
    force=False,
    rand_low=-100,
    rang_high=100,
    log=True,
):
    if filename is None:
        filename = os.path.abspath(
            f"int_dataset-{nrows},{ncols},{rand_low},{rang_high},{multiline_rows}.csv"
        )

    print("generating", filename)

    if os.path.exists(filename) and not force:
        return filename

    data = {
        f"col{i}": np.concatenate(
            [
                np.concatenate(
                    [
                        ["some\nvery very very\nlong string\nwith many multilines"],
                        np.random.randint(rand_low, rang_high, 9),
                    ]
                )
                for _ in np.arange(ncols // 10)
            ]
        )
        if (i % 10 == 0 and multiline_rows)
        else list(np.random.randint(rand_low, rang_high, nrows // 10)) * 10
        for i in np.arange(ncols)
    }
    print("dict generated!")
    df = pandas.DataFrame(data)
    print("dataframe created!")
    df.to_csv(filename)
    print("csv ready!")
    return filename


if __name__ == "__main__":
    import pandas
    import modin.pandas as pd

    fname = generate_data_file(1_000_000)

    df = pd.read_csv(fname)
    cols_to_grab = ["col1", "col64", "col100", "col255"]
    t1 = timer()
    result = [
        df.loc[:, [col]].squeeze(axis=1).values
        for col in cols_to_grab
    ] + [list(df.index)]
    print("grabbing cols by one", timer() - t1)

    t1 = timer()
    pd_df = np.concatenate([df[cols_to_grab].values.T, [df.index]])
    print("grabbing all", timer() - t1)

```


</details>

my suggestion will be to rewrite this in something like this:
```python
new_idx_lvl_arrays = np.concatenate([self._df[self._by.columns].values.T, [result.index]])
```

Reply:
Good point, thanks. It was applied with minor change: `.., [list(result.index)]`.

Reply:
hm, I've tested this code locally and everything was okay, do we really need to convert index to list?
```python
res1 = pandas.MultiIndex.from_arrays(
    [df.loc[:, [col]].squeeze(axis=1).values for col in cols_to_grab]
    + [list(df.index)]
)

res2 = pandas.MultiIndex.from_arrays(
    np.concatenate([df[cols_to_grab].values.T, [df.index]])
)

assert res1.equals(res2)
```

Reply:
This will work without `list()` when `type(df.index)` is not `DatetimeIndex`.

Reply:
And this code branch works only when `type(df.index)` is datetime-like.